clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,390 @@ package common
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, nil, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addF := func() {
+				w.handleObject(c, nil, obj)
+			}
+			w.queue.Add(&addF)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			updateF := func() {
+				w.handleObject(c, oldObj, newObj)
+			}
+			w.queue.Add(&updateF)
+		},
+		DeleteFunc: func(obj interface{}) {
+			deleteF := func() {
+				w.handleObject(c, obj, nil)
+			}
+			w.queue.Add(&deleteF)
+		},
+	})
+
+	for i := 0; i < waitForControlledPodsWorkers; i++ {
+		w.workerGroup.Start(w.worker)
+	}
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) worker() {
+	for {
+		f, stop := w.queue.Get()
+		if stop {
+			return
+		}
+		(*f.(*func()))()
+		w.queue.Done(f)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, syncTimeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", w)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", w)
+	}
+	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion(c)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+
+	cond := func() (bool, error) {
+		return w.opResourceVersion >= maxResourceVersion, nil
+	}
+	if err := wait.Poll(checkControlledPodsInterval, syncTimeout, cond); err != nil {
+		return fmt.Errorf("timed out while waiting for controlled pods")
+	}
+
+	w.handlingGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	numberRunning := 0
+	for _, checker := range w.checkerMap {
+		if status := checker.getStatus(); status {
+			numberRunning++
+		}
+	}
+	if desiredCount != numberRunning {
+		glog.Errorf("%s: incorrect objects number: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+		return fmt.Errorf("incorrect objects number")
+	}
+
+	glog.Infof("%s: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.workerGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, oldObj, newObj interface{}) {
+	var oldRuntimeObj runtime.Object
+	var newRuntimeObj runtime.Object
+	var ok bool
+	oldRuntimeObj, ok = oldObj.(runtime.Object)
+	if oldObj != nil && !ok {
+		return
+	}
+	newRuntimeObj, ok = newObj.(runtime.Object)
+	if newRuntimeObj != nil && !ok {
+		return
```

Comment:
don't fail silently

> That's true. However, I expect that all small errors will have impact on final measurement check (e.g. failing to check object deletion will result on failing object count check), so all in all whole measurement will fail.

Not necessary. For this particular case, imagine that RC was successfully created and thus it has a checker with final status.
But then if you want to update it and you fail e.g. here, it simply won't have any effect. - we won't do nothing, but the measurement itself will succeed.

As i said, a bunch of errors here, should cause test to fail (though not immediately).

Reply:
I've added `glog.Errorf`.

> But then if you want to update it and you fail e.g. here, it simply won't have any effect. - we won't do nothing, but the measurement itself will succeed.

I see two possible scenarios:
- There is unrecognizable object - the cast fails or some of necessary fields are not accessible. I think it should be log - it something unexpected, but that's definitely not a RC (or other controlling pods object), so it shouldn't affect wait for controlled pods measurement.
- Object is recognized, therefore checker is created. If checker fails, it will be visible because of he status. The measurement will fail.

IMHO, this is handled properly.

Anyway, if you're suggesting that errors should be handle in the different way, maybe we want to create issue? TBH I think we want to moved this topic to the different PR to iron out the proper solution.

Reply:
Being unable to cast means some logical error in the code - we are doing something really wrong then. This should fail the test.

I'm fine with opening a separate issue for that and solving it separately (it may be the case fo different measurements and other parts of code). But this needs to be fixed before we switch to cluster-loader for running our tests.

Reply:
> But this needs to be fixed before we switch to cluster-loader for running our tests.

I agree. I created #276 - this will allow us to track error fixing status for all measurements.